daemon/entrypoint.sh

```diff
@@ -0,0 +1,327 @@
+       else
+          OSD_J=/var/lib/ceph/osd/${CLUSTER}-${OSD_ID}/journal
+       fi
+    fi
+
+    # Check to see if our OSD has been initialized
+    if [ ! -e /var/lib/ceph/osd/${CLUSTER}-${OSD_ID}/keyring ]; then
+      # Create OSD key and file structure
+      ceph-osd -i $OSD_ID --mkfs --mkjournal --osd-journal ${OSD_J}
+
+      if [ ! -e /var/lib/ceph/bootstrap-osd/ceph.keyring ]; then
+        echo "ERROR- /var/lib/ceph/bootstrap-osd/ceph.keyring must exist. You can extract it from your current monitor by running 'ceph auth get client.bootstrap-osd -o /var/lib/ceph/bootstrap-osd/ceph.keyring'"
+        exit 1
+      fi
+
+      timeout 10 --cluster ${CLUSTER} ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring health || exit 1
+
+      # Generate the OSD key
+      ceph --cluster ${CLUSTER} --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring auth get-or-create osd.${OSD_ID} osd 'allow *' mon 'allow profile osd' -o /var/lib/ceph/osd/${CLUSTER}-${OSD_ID}/keyring
+
+      # Add the OSD to the CRUSH map
+      if [ ! -n "${HOSTNAME}" ]; then
+        echo "HOSTNAME not set; cannot add OSD to CRUSH map"
+        exit 1
+      fi
+      ceph --name=osd.${OSD_ID} --keyring=/var/lib/ceph/osd/${CLUSTER}-${OSD_ID}/keyring osd crush create-or-move -- ${OSD_ID} ${OSD_WEIGHT} root=default host=${HOSTNAME}
+    fi
+
+    mkdir -p /etc/service/ceph-${OSD_ID}
+    cat >/etc/service/ceph-${OSD_ID}/run <<EOF
+#!/bin/bash
+echo "store-daemon: starting daemon on ${HOSTNAME}..."
+exec ceph-osd -f -d -i ${OSD_ID} --osd-journal ${OSD_J} -k /var/lib/ceph/osd/ceph-${OSD_ID}/keyring
+EOF
+    chmod +x /etc/service/ceph-${OSD_ID}/run
+  done
+
+read
+
+
+#################
+# OSD_CEPH_DISK #
+#################
+
+elif [[ "$CEPH_DAEMON" = "OSD_CEPH_DISK" ]]; then
+
+  ceph_config_check
+
+  if [[ -z "${OSD_DEVICE}" ]];then
+    echo "ERROR- You must provide a device to build your OSD ie: /dev/sdb"
+    exit 1
+  fi
+
+  if [ ! -e /var/lib/ceph/bootstrap-osd/ceph.keyring ]; then
+    echo "ERROR- /var/lib/ceph/bootstrap-ods/ceph.keyring must exist. You can extract it from your current monitor by running 'ceph auth get client.bootstrap-ods -o /var/lib/ceph/bootstrap-ods/ceph.keyring'"
+    exit 1
+  fi
+
+  timeout 10 --cluster ${CLUSTER} ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring health || exit 1
+
+  mkdir -p /var/lib/ceph/osd
+
+  # TODO:
+  # -  add device format check (make sure only one device is passed
+
+  if [[ "$(sudo parted --script ${OSD_DEVICE} print | egrep '^ 1.*ceph data')" ]]; then
+    echo "ERROR- It looks like this device is an OSD, use OSD_FORCE_ZAP to use this device"
+    exit 1
+  elif [[ "$(sudo parted --script ${OSD_DEVICE} print | egrep '^ 1.*ceph data')" && ${OSD_FORCE_ZAP} -eq "1" ]]; then
+    ceph-disk -v zap ${OSD_DEVICE}
+  fi
+
+  if [[ ! -z "${OSD_JOURNAL}" ]]; then
+    ceph-disk -v prepare ${OSD_DEVICE}:${OSD_JOURNAL}
```

Comment:
What does ceph-disk do in this case? I can't recall what it supports. Do partitions need to be precreated? Will it format the journal if it doesn't exist?